pkg/shared/accountmgmtutil/ams.go

```diff
@@ -68,144 +70,180 @@ func fetchOrgQuotaCost(ctx context.Context, conn connection.Connection) (*amscli
 	}
 
-	quotaCostGet, err := fetchOrgQuotaCost(ctx, conn)
+	marketplaceQuota, err := pickMarketplaceQuota(f, marketplaceQuotas, marketplace)
 	if err != nil {
 		return nil, err
 	}
+	return marketplaceQuota, nil
 
-	for _, quota := range quotaCostGet.GetItems() {
-		if len(quota.GetCloudAccounts()) > 0 {
-			for _, cloudAccount := range quota.GetCloudAccounts() {
-				if marketplace != "" {
-					if cloudAccount.GetCloudProviderId() == marketplace {
-						marketplaceAcctIDs = append(marketplaceAcctIDs, cloudAccount.GetCloudAccountId())
-					}
-				} else {
-					marketplaceAcctIDs = append(marketplaceAcctIDs, cloudAccount.GetCloudAccountId())
+}
+
+func pickMarketplaceQuota(f *factory.Factory, marketplaceQuotas []QuotaSpec, marketplace MarketplaceInfo) (*QuotaSpec, error) {
+
+	matchedQuotas := []QuotaSpec{}
+
+	for _, quota := range marketplaceQuotas {
+		cloudAccounts := *quota.CloudAccounts
+		for _, cloudAccount := range cloudAccounts {
+			if marketplace.Provider != "" && marketplace.CloudAccountID != "" {
+				if *cloudAccount.CloudProviderId == marketplace.Provider && *cloudAccount.CloudAccountId == marketplace.CloudAccountID {
+					matchedQuotas = append(matchedQuotas, quota)
+				}
+			} else if marketplace.Provider != "" || marketplace.CloudAccountID != "" {
+				if *cloudAccount.CloudProviderId == marketplace.Provider || *cloudAccount.CloudAccountId == marketplace.CloudAccountID {
+					matchedQuotas = append(matchedQuotas, quota)
 				}
 			}
 		}
 	}
 
-	return unique(marketplaceAcctIDs), err
+	if len(matchedQuotas) == 0 {
+		return nil, f.Localizer.MustLocalizeError("kafka.create.quota.error.cloudAccountNotFound")
+	}
+
+	return &matchedQuotas[0], nil
 }
 
-func GetValidMarketplaces(ctx context.Context, connectionFunc factory.ConnectionFunc) (marketplaces []string, err error) {
+func pickCloudAccount(f *factory.Factory, cloudAccounts *[]amsclient.CloudAccount, market MarketplaceInfo) (*[]amsclient.CloudAccount, error) {
 
-	conn, err := connectionFunc(connection.DefaultConfigSkipMasAuth)
-	if err != nil {
-		return nil, err
+	if len(*cloudAccounts) == 1 {
+		return cloudAccounts, nil
 	}
 
-	quotaCostGet, err := fetchOrgQuotaCost(ctx, conn)
-	if err != nil {
-		return nil, err
+	if len(*cloudAccounts) > 2 && market.Provider == "" && market.CloudAccountID == "" {
+		return nil, f.Localizer.MustLocalizeError("kafka.create.quota.error.multipleCloudAccounts")
 	}
 
-	for _, quota := range quotaCostGet.GetItems() {
-		if len(quota.GetCloudAccounts()) > 0 {
-			for _, cloudAccount := range quota.GetCloudAccounts() {
-				marketplaces = append(marketplaces, cloudAccount.GetCloudProviderId())
+	var matchedAccounts []amsclient.CloudAccount
+
+	for _, cloudAccount := range *cloudAccounts {
+		if market.Provider != "" || market.CloudAccountID != "" {
+			if *cloudAccount.CloudProviderId == market.Provider || *cloudAccount.CloudAccountId == market.CloudAccountID {
```

Comment:
Question I have:
 Do you think we should support people specifying only market.Provider?

Maybe we can just:
```suggestion
			if *cloudAccount.CloudProviderId == market.Provider && *cloudAccount.CloudAccountId == market.CloudAccountID {
```

To keep this convoluted logic little bit simpler?

Reply:
> Do you think we should support people specifying only market.Provider?

I will agree not allowing marketplace provider alone can make the logic bit simpler. I will be removing it.